File: compiler/src/feedback/items/SyntaxFeedback.swift
func registerSyntaxFeedback(into context: FeedbackContext) {
    context.register(
        id: "syntax-error",
        query: Queries.fact(ParseError.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, _, error in
        writer.write(error.message)

        if let committed = error.committed {
            writer.write(" ")
            writer.write(committed)
        }

        writer.write(".")

        if let reason = error.reason {
            writer.writeBreak()
            writer.write(reason)
        }

        writer.writeBreak()
        writer.write("Check your spelling.")
    }

    context.register(
        id: "unused-block",
        query: { (db, node, body: (Node) -> Void) in Queries.unusedBlock(db, node) { body(node) } },
        rank: { _ in .syntax },
        location: { node in (primary: node, secondary: []) },
    ) { writer, node in
        writer.write(node)
        writer.write(" is on its own line and will never run.")
        writer.writeBreak()
        writer.write("Did you mean to put the opening ")
        writer.write(code: "{")
        writer.write(" on the line above?")
    }
}

File: compiler/feedback/src/items/syntax_feedback.rs
use crate::{FeedbackCtx, FeedbackRank};
use wipple_queries::{syntax_error, unused_block};

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("syntax-error")
        .query(syntax_error)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, _, error| {
            writer.string(&error.message);

            if let Some(committed) = &error.committed {
                writer.string(" ");
                writer.string(committed);
            }

            writer.string(".");

            if let Some(reason) = &error.reason {
                writer.line_break();
                writer.string(reason);
            }

            writer.line_break();
            writer.string("Check your spelling.");
        })
        .register();

    ctx.feedback("unused-block")
        .query(unused_block)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" is on its own line and will never run.");
            writer.line_break();
            writer.string("Did you mean to put the opening ");
            writer.code("{");
            writer.string(" on the line above?");
        })
        .register();
}
